greenwave/policies.py

```diff
@@ -471,35 +495,34 @@ def to_json(self):
         }
 
 
-def _summarize_answers_without_errored(answers):
-    failure_count = sum(
-        1 for answer in answers
-        if isinstance(answer, RuleNotSatisfied)
-    )
-    missing_count = sum(
-        1 for answer in answers
-        if isinstance(answer, (TestResultIncomplete, TestResultMissing))
-    )
-
-    # Missing results are also failures but we will distinguish between those
-    # two in summary message.
-    failure_count -= missing_count
-
-    if failure_count and missing_count:
-        return '{} of {} required tests failed, {} result{} missing'.format(
-            failure_count, len(answers), missing_count, 's' if missing_count > 1 else '')
-
-    if failure_count > 0:
-        return '{} of {} required tests failed'.format(failure_count, len(answers))
-
-    if missing_count > 0:
-        return '{} of {} required test results missing'.format(missing_count, len(answers))
-
-    if all(answer.is_satisfied for answer in answers):
-        return 'All required tests passed'
+class _Summary(object):
+    """
+    Internal class using for generating the result summary.
+    """
+    def __init__(self):
+        self.test_msgs = defaultdict(int)
+        self.non_test_msgs = defaultdict(int)
+
+    def to_text(self, test_count):
+        msgstr = ""
+        if self.non_test_msgs:
+            msgstr = ", ".join(
+                self.sformat(f"{{num}} {msg}", num) for (msg, num) in self.non_test_msgs.items()
+            )
+        if self.test_msgs:
+            addmsg = self.sformat("Of {num} required test{s}, ", test_count)
+            addmsg += ", ".join(
+                self.sformat(f"{{num}} {msg}", num) for (msg, num) in self.test_msgs.items()
```

Comment:
Does something ensure these messages are consistent? E.g. "failed results" are always mentioned before "incomplete", and "incomplete" before "missing" etc.

In recent Python 3.7 and up, the dicts items are guaranteed to be ordered by the insertion time (I assume this also applies to `defaultdict`) which I think would mean that the messages would depend on what kinds of results (failed/passed/missing) are encountered first.